// @sanity/runtime-cli/dist/commands/blueprints/info.js
import { Flags } from '@oclif/core';
import { DeployedBlueprintCommand } from '../../baseCommands.js';
import { blueprintInfoCore } from '../../cores/blueprints/info.js';
export default class InfoCommand extends DeployedBlueprintCommand {
    static description = 'Show information about a Blueprint deployment';
    static examples = [
        '<%= config.bin %> <%= command.id %>',
        // LAUNCH LIMIT: 1 Stack per Project - do not allow Stack ID to be set
        // '<%= config.bin %> <%= command.id %> --id ST-a1b2c3',
    ];
    static flags = {
        id: Flags.string({
            description: 'Stack ID to show info for (defaults to current stack)',
            hidden: true, // LAUNCH LIMIT: 1 Stack per Project
        }),
    };
    async run() {
        const { success, error } = await blueprintInfoCore({
            bin: this.config.bin,
            log: (message) => this.log(message),
            auth: this.auth,
            stackId: this.stackId,
            deployedStack: this.deployedStack,
            flags: this.flags,
        });
        if (!success)
            this.error(error);
    }
}
