// @sanity/runtime-cli/dist/commands/blueprints/stacks.js
import { Flags } from '@oclif/core';
import { BlueprintCommand } from '../../baseCommands.js';
import { blueprintStacksCore } from '../../cores/blueprints/stacks.js';
export default class StacksCommand extends BlueprintCommand {
    // LAUNCH LIMIT: 1 Stack per Project - hide stacks command
    static hidden = true;
    static description = 'List all Blueprint stacks';
    static examples = [
        '<%= config.bin %> <%= command.id %>',
        '<%= config.bin %> <%= command.id %> --projectId a1b2c3',
    ];
    static flags = {
        projectId: Flags.string({
            description: 'Project ID to show stacks for',
            required: false,
        }),
    };
    async run() {
        const { success, error } = await blueprintStacksCore({
            bin: this.config.bin,
            log: (message) => this.log(message),
            token: this.sanityToken,
            blueprint: this.blueprint,
            flags: this.flags,
        });
        if (!success)
            this.error(error);
    }
}
