// @portabletext/editor/lib/_chunks-cjs/util.selection-point-to-block-offset.cjs
"use strict";
var util_childSelectionPointToBlockOffset = require("./util.child-selection-point-to-block-offset.cjs"), util_sliceBlocks = require("./util.slice-blocks.cjs");
function getSelectionEndPoint(selection) {
  return selection ? selection.backward ? selection.anchor : selection.focus : null;
}
function getSelectionStartPoint(selection) {
  return selection ? selection.backward ? selection.focus : selection.anchor : null;
}
function isSelectionCollapsed(selection) {
  return selection ? JSON.stringify(selection.anchor.path) === JSON.stringify(selection.focus.path) && selection.anchor.offset === selection.focus.offset : !1;
}
function selectionPointToBlockOffset({
  context,
  selectionPoint
}) {
  return selectionPoint.path.length === 1 && util_sliceBlocks.isKeyedSegment(selectionPoint.path[0]) ? {
    path: [{
      _key: selectionPoint.path[0]._key
    }],
    offset: selectionPoint.offset
  } : util_childSelectionPointToBlockOffset.childSelectionPointToBlockOffset({
    context,
    selectionPoint
  });
}
exports.getSelectionEndPoint = getSelectionEndPoint;
exports.getSelectionStartPoint = getSelectionStartPoint;
exports.isSelectionCollapsed = isSelectionCollapsed;
exports.selectionPointToBlockOffset = selectionPointToBlockOffset;
